external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala

```diff
@@ -118,6 +123,7 @@ class DirectKafkaInputDStream[
       context.sparkContext, kafkaParams, currentOffsets, untilOffsets, messageHandler)
 
     currentOffsets = untilOffsets.map(kv => kv._1 -> kv._2.offset)
+    offsetMap += ((validTime, currentOffsets))
```

Comment:
I agree that offsetMap has redundant info, but unless checkpointing is turned on, prior offset ranges aren't tracked.

As for the synchronization, if multiple threads aren't touching offsetMap, why do you have the synchronized block in the listener?